Local notification fixes

ray-kraesig opened the pull request:
* Correctly hook up to `react-native-notifications`. (Fixes #3672.)
* Verify (including by typechecking) the contents of APNs notification data.
* Remove some muddled [_sic_], now-unnecessary rectification code.

N.B.: The first commit here is a duplicate of the current #3906, and should be considered there.

ray-kraesig commented:
:warning: As this is APNs notifications code, it has not yet been properly tested.

gnprice commented on src/notification/index.js:
```diff
@@ -233,8 +228,24 @@ export class NotificationListener {
+      // data it returns from `getData` is unvalidated -- it comes almost
+      // straight off the wire from the server.
+      this.listen('notificationOpened', (note: { getData(): JSONableDict }) => {
+        const data = verifyIosNotificationData(note);
```
Hmm, something doesn't seem right in these types! We're passing `verifyIosNotificationData` the thing that has a `getData()` method, rather than that method's return value.

I'm actually puzzled here that Flow doesn't give an error.

ray-kraesig replied:
> I'm actually puzzled here that Flow doesn't give an error.

This is, alas, a known bug. See comment in `jsonable.js`.

gnprice replied:
Ah, I see. It's because `JSONableDict`, which is the type of that parameter, is an inexact object type. (Which in turn it has to be because of that Flow bug fixed in v0.111.)

gnprice commented on src/notification/verifyIosNotificationData.js:
```diff
@@ -0,0 +1,232 @@
+ *
+ * @returns A `Notification` on success; `undefined` on failure.
+ */
+export default (data: JSONableDict): Notification | void => {
```
I think the name actually undersells what this module does! This *parses* the notification data. Like any good parser (well, any good parser for normal circumstances), it indeed validates the format of its input along the way. But the main verb for what it does is "parse", because it actually returns the pieces of the data the application needs, in a well-structured form that requires little or no downstream work to make use of.

I think simply s/verify/parse/ would make a good name.

ray-kraesig replied:
It did stop being a mere verifier somewhere along the line; but "parse" implies a specific, and very different, class of transformation (string-to-tree). How does "extract" strike you?

gnprice replied:
WFM!

gnprice commented on src/notification/types.js:
```diff
@@ -0,0 +1,27 @@
+ *
+ *  * On iOS, these objects are translated directly, field by field, from
+ *    the blobs of key/value pairs sent by the Zulip server in the
+ *    "payload".  The set of fields therefore varies by server version.
```
I think at the tip of the branch this jsdoc can be updated, to be made happier :grinning: 

(That's assuming that now all values we describe as `Notification`, in the iOS case, come from the new validating parser. Which I believe is a consequence of this branch, but haven't confirmed.)

ray-kraesig commented on src/notification/index.js:
```diff
@@ -202,25 +174,48 @@ export class NotificationListener {
+      }
+
+      // A non-null device token should be a string of hex digits.
+      if (typeof deviceToken !== 'string' || !deviceToken.match(/^[0-9a-fA-F]$/)) {
```
I was mostly looking at APNs, but I thought I'd seen either a hex-string FCM token, or someplace saying that FCM tokens were also hex-strings.

I'd go hunt it down, but your second point makes it entirely irrelevant. Out it goes!

ray-kraesig commented on src/notification/verifyIosNotificationData.js:
```diff
@@ -0,0 +1,232 @@
+ *
+ * Unlike the definition of `JSONableDict`, this permits the possiblility of
+ * `void` elements. This is a workaround for Flow's by-design soundness hole
+ * involving indexer elements.
```
Alas, [the following is documented as deliberate](https://flow.org/en/docs/types/objects/#toc-objects-as-maps):

```flow
const indexedMap: { +[string]: number => number } = { sqrt: Math.sqrt };
const value: number = indexedMap['srqt'](42);  // kaboom!
```

gnprice replied:
Got it, thanks for the link. Here's the key bit:

> When an object type has an indexer property, property accesses are assumed to have the annotated type, even if the object does not have a value in that slot at runtime. It is the programmer’s responsibility to ensure the access is safe, as with arrays.

I think the main thing this comment needs is just to be specific about what you mean: it's the fact that `data.foo` on a JSONableDict will be treated as if the `foo` property is present. Even after reading this whole file, I had no idea what fact about Flow you had in mind or what this type was for.

ray-kraesig replied:
Sorry – I've run into it so many times that it just feels like background information about Flow.

(More verbose explanation given next to a new `JSONableInput` type in `jsonable.js`.)

ray-kraesig commented on src/notification/index.js:
```diff
@@ -189,9 +189,10 @@ export class NotificationListener {
     }
-    const subscription = DeviceEventEmitter.addListener(name, handler);
-    this.unsubs.push(() => subscription.remove());
   }
```
> I'm also not sure we need a fancy acronymic name for our Android-native notifications code :slightly_smiling_face:

Editing failure; I thought I got rid of it. :disappointed: Fixed.

ray-kraesig commented on src/notification/index.js:
```diff
@@ -260,8 +257,24 @@ export class NotificationListener {
+        const data = extractNotificationData(note);
+        if (data) {
+          this.handleNotificationOpen(data);
+        }
```
No, there's a very big change here – messages from the Android side are no longer being run through `extractNotificationData`.

On the iOS side, though, it is supposed to be `[nfc]` (though that could be explained better). The conditional in question is effectively lifted from `narrowToNotification`, which is the external destination of this value; it cannot be removed from this commit without replacing the `$FlowFixMe` deleted from `handleNotificationOpen`.

Part of the point of this change is to make `handleNotificationOpen` properly typed, internally and externally, and to take precisely a `Notification` rather than a `NotificationMuddle`.

gnprice replied:
> No, there's a very big change here – messages from the Android side are no longer being run through `extractNotificationData`.

Hmm, but I think they already weren't:
```
-  handleNotificationOpen = (notification: NotificationMuddle) => {
-    const data: Notification =
-      // $FlowFixMe clarify notification types
-      Platform.OS === 'ios' ? extractNotificationData(notification) : notification;
-    this.dispatch(narrowToNotification(data));
```
In the Android case, that doesn't call `extractNotificationData`.

gnprice replied:
> The conditional in question is effectively lifted from `narrowToNotification`, which is the external destination of this value; it cannot be removed from this commit without replacing the `$FlowFixMe` deleted from `handleNotificationOpen`.

Deleting fixmes is a great thing!

Even better, though, is deleting them in a minimal change, before and/or after NFC commits that make related changes. It looks like that should be possible here.

ray-kraesig replied:
> In the Android case, that doesn't call `extractNotificationData`.

:facepalm: Right. Somehow completely missed that on my first *and* second pass. Will adjust.

ray-kraesig commented on src/notification/verifyIosNotificationData.js:
```diff
@@ -0,0 +1,232 @@
+  // hands us the rest as "data". Pretty much any iOS notifications library
+  // should do the same, but we don't rely on that.
+
+  const data: PartialJSONableDict = (() => {
```
To prevent accidentally hitting the indexer-property soundness hole. Flow is indeed happy if you do that, but Flow will remain happy if you do that and then make a lot of accesses to `data` that won't work.

(This could stand to be in a comment.)

gnprice replied:
Could you make a version of `PartialJSONableDict` that allows for `void`s to appear recursively? (Perhaps in `jsonable.js` so it's next to the definition of `JSONable`, with which it shares some facts.) Then it should be possible to mark just one thing at the outside as "partial", and Flow would catch if we dropped the partialness somewhere along the way.

That one spot at the outside would then be a good spot for a short comment mentioning why it's needed.

ray-kraesig replied:
> That one spot at the outside would then be a good spot for a short comment mentioning why it's needed.

That deserves a long comment, really. (Done.)

ray-kraesig commented on src/notification/verifyIosNotificationData.js:
```diff
@@ -0,0 +1,232 @@
+  //
+  // [0] GitHub commit 410ee44eb6e40bac4099d1851d949533026fe4b3, imported from
+  //     an older commit 6c54fe44a82c5796268e56d3f5577bf4cfc8163a.
+
```
Yeah, not intentional. :disappointed:

ray-kraesig commented on src/notification/verifyIosNotificationData.js:
```diff
@@ -0,0 +1,232 @@
+  if (typeof recipient_type !== 'string') {
+    throw err('invalid');
+  }
+  if (!RECIPIENT_TYPES[recipient_type]) {
```
At one point I was trying to make Flow narrow `recipient_type` to `'stream' | 'private'`. (It didn't work out; this was a relic.)

gnprice replied:
Ah, makes sense! That would indeed be nice.

gnprice commented:
Pushed the bugfix commit, as 400864361. Thanks!

ray-kraesig commented on src/notification/verifyIosNotificationData.js:
```diff
@@ -0,0 +1,232 @@
+        recipient_type: 'private',
+
+        // added 1.7.0-2360-g693a9a5e70, release 1.8.0+
+        // present only on huddles (multi-user PMs)
```
You say that, but we've got a fair number of occurrences of "huddles" in the client code right now: `UnreadHuddlesState`, `addItemsToHuddleArray`, `getUnreadByHuddles`, etc.

(Altered anyway.)

gnprice replied:
Yeah. Those are all within the `state.unread` reducer code, though, and they're all due to a single such slip-up in the API, in the `unread_msgs` part of the initial data (on `/register`, or in `page_params` for the webapp). That data structure (a) distinguishes group PMs from 1:1 PMs, and (b) uses the term "huddle" for the former, both of which we've otherwise managed to avoid in the API.

ray-kraesig commented on src/notification/__tests__/notification-test.js:
```diff
@@ -33,11 +44,11 @@ describe('getNarrowFromNotificationData', () => {
+      [4, makeUser({ name: 'john' })],
     ]);
-    const expectedNarrow = groupNarrow(['me@example.com', 'mark@example.com', 'john@example.com']);
+    const expectedNarrow = groupNarrow(['me@example.org', 'mark@example.org', 'john@example.org']);
```
I'm not a great fan of that form in particular (I think we want a make-users-by-ID helper function somewhere), but it's certainly better than what I wrote, for all the reasons you've given.

I'll poke at it a bit and see if I can't do better without significantly bloating the PR. If I can't, well, follow-up work. :shrug:

gnprice replied:
New version lgtm, thanks!

ray-kraesig commented:
Ready for re-review.

gnprice commented on src/utils/jsonable.js:
```diff
@@ -37,3 +45,56 @@ export type JSONable =
+ *
+ * As JSONableInput <: JSONable, JSON values should generally be passed around
+ * as `JSONable` until and unless they're actually being validated; they can
+ * then be locally downgraded to `JSONableInput`.
```
Rather JSONable <: JSONableInput, right? I read that operator as "is a subtype of".

And then... "downgraded" sure is an accurate characterization (as it's e.g. a less informative type), but potentially a confusing word given the mixed metaphor with the type hierarchy :upside_down_face: . Maybe just "be locally cast to"? Or s/cast/upcast/.

ray-kraesig replied:
An upcast is a downgrade. There's nothing confusing there.  ⍩⃝  

Fixed and amended.

gnprice commented on src/notification/extract.js:
```diff
@@ -0,0 +1,223 @@
+
+/** JSONableInput dictionary. */
+// eslint-disable-next-line flowtype/type-id-match
+type JSONableInputDict = { +[string]: JSONableInput | void };
```
You also add one of these in `jsonable`. That seems like a fine place -- import from there?

ray-kraesig replied:
... I swear I *did* that. (Fixed.)

gnprice commented:
Thanks @ray-kraesig for these revisions! These changes generally look great.

Some small comments above.

Then for the `handleDeviceToken` / `validateToken` changes in the first two commits, I have some thoughts on the code -- but also ISTR we looked at it in person sometime in the past couple of weeks, and you had a revised version that I liked better. Would you update this PR branch to use that version?

ray-kraesig commented:
> Then for the `handleDeviceToken` / `validateToken` changes in the first two commits, I have some thoughts on the code -- but also ISTR we looked at it in person sometime in the past couple of weeks, and you had a revised version that I liked better. Would you update this PR branch to use that version?

Technically, no – because I can't find it. :upside_down_face: Rewritten, by simplification.

gnprice commented:
Thanks for the revisions! I like the new version of `handleDeviceToken`. :slightly_smiling_face: 

Merged, with a few small edits, and an added jsdoc commit on top.

ray-kraesig commented:
> Technically, no – because I can't find it. :upside_down_face: Rewritten and simplified.

[Found it.](https://github.com/zulip/zulip-mobile/pull/3906) :sweat_smile:
